docs/rfc/0000-cli-infrastructure.adoc

```diff
@@ -0,0 +1,237 @@
+iteration of the core set of subcommands.
+
+== One `rad` to Rule Them All
+
+[quote]
+-------
+One ring to rule them all, one ring to find them,
+One ring to bring them all, and in the darkness bind them;
+-------
+-- J.R.R. Tolkein, Fellowship of the Ring
+
+The integral part of this proposal is that there will be one binary,
+`rad`, that will be the foundation for all other subcommands. It will
+define the global parameters and how to extend it.
+
+We will discuss these in the following two sections.
+
+== The Fellowship of the `rad`
+
+[quoute]
+--------
+“I can’t carry it for you, but I can carry you.” – Sam Gamgee
+--------
+-- J.R.R. Tolkien, The Return of the King
+
+To allow the ecosystem of commands to be extended by ourselves and
+others in the community it will be necessary to extend the `rad`
+command with further subcommands. The details of this are outlined in
+RFC-682#CLI, however, we will give a quick overview here for the sake
+of the reader.
+
+Taking inspiration from tools such as `git` and `cargo`, we propose
+that the `rad` binary can be extended by placing another
+binary on the users `$PATH` (or `%PATH%` on Windows). This
+binary MUST follow the naming convention of `rad-*`. To reiterate
+here, these new commands MUST NOT override the core set outlined here
+(nor in further RFCs).
+
+== Global Parameters
+
+Many subcommands will likely want to share a set of options. These
+options, dubbed as "global parameters" here, will be a defined set --
+exported from `radicle-link` for re-use -- that the base `rad` command
+will always receive and pass onto the subcommand being called.
+
+Thus, all subcommands SHALL receive these global parameters as part of
+their interface. We list these parameters here, but it is NOT an
+exhaustive list:
+
+* `verbose`
+* `quiet`
+* `profile := active (default) | <profile-id>`
+
+== Online & Offline
+
+In RFC-682#Process-Orchestration, it was decided that the peer-to-peer
+damon would not be a long-running process. The terminology we will use
+here is that the daemon is *online* if it is running and connected to
+the network. Otherwise, if it is not running then it is *offline*.
+
+This disctinction becomes important when thinking about the
+functionality of `rad` subcommands. The majority of subcommands that
+interact with the storage can be done while the daemon is offline. The
+only action the author can currently think of — that is online — is
+replication of a URN from the network.
+
+This leaves us in a predicament of thinking about how multiple Radicle
+applictions may interact with each other. For example, if Upstream is
+running and I create a project via the CLI, then a good user
+experience is that the Upstream is updated with the new project
+(without refreshing). In RFC-682, it is alluded that a PubSub
+component could solve this, however, that requires the daemon to be
+online. The author prefers not to specify the PubSub component
```

Comment:
Could be I miss an angle here, but the PubSub as described in 682 is dependent on an "online" p2p node. In contrast it specifically recommends a substrate which would be a system service. Or is your meaning here that the node needs to be running to observe the message from the PubSub?

Reply:
> In contrast it specifically recommends a substrate which would be a system service.

I'm not sure what you mean here. Could you maybe rephrase it?

> Or is your meaning here that the node needs to be running to observe the message from the PubSub?

Ya, exactly. In the example, if I created a project via the CLI and I also have Upstream running (without the daemon running), then there's no way for Upstream to know that an event happened (the project being created).

Reply:
If upstream is running, it should connect to the pubsubbub, onto which the CLI places a message after creating the project. No daemon involved

Reply:
I thought we hadn't landed on whether pubsubflub was a standalone service or if it was only up along with the daemon? If it's always available then I guess this section is pointless :upside_down_face: But it might be useful to have a section on events being emitted, what do yis think?

Reply:
I think we weren't sure what the right tool is (D-Bus being rather baroque). But I'm fairly sure that we want some sort of standalone thing, because it can become confusing quickly to see what actually happened. For example:

Say you unplugged your WiFi cable, so whenever the daemon comes up, it won't be able to connect to the network. Now say you have Upstream running, but you do something using another program (eg. push some changes using `git`). What should happen?

One option is that the daemon keeps running (because Upstream is connected to it), gets notified by `git` that something happened, and then sends `updated-refs ...` to Upstream, followed by `not-connected`. The other option is that the daemon exits when it can't connect (a few times). It will miss the notification from `git`, and Upstream will just wait in vain for some kind of event which tells it that something was replicated to the network.

Both seem pretty terrible (as are most "event driven" systems) -- you never know if and when the next message arrives, and "no message" can carry information. We know, however, that we need to record the modification of the repo _either way_, such that the daemon can pick it up once it gains connectivity.

The missing piece here -- and this is not the first time I'm saying this -- is that we can't have "offline first" and "it behaves like GitHub" at the same time. In other words, the act of synchronising the local state with the network must be controllable by the user (and is genuinely an "online" operation, btw). 

So in the example: 

If Upstream is running, receives a notification that the repo state has changed, and decides to pretend "it behaves like GitHub", it is _Upstream's_ responsibility to call a `sync` RPC on the daemon, and synchronously wait for the result. That `sync` RPC obviously needs to be idempotent, and it should probably be possible to retrieve information about when and what the last sync did.

If Upstream is _not_ running, things are slightly more complicated, because we can't have `git push` to also atomically `sync` (the refs must have been written already, and there's no atomic rollback). We can do this non-atomically, though, and just funnel back the result to `git` through the sideband. A `rad git push` wrapper would allow to make it optional (by either passing `--sync` or not).

So in summary, there are zero assumptions one can have about the availability of other processes in the service mesh. Therefore, events are always ephemeral, and it is up to the receiver to act upon them, not the sender. Actions which return a result must be synchronous RPCs. 

A consequence of this is that there must be ways to synchronously inspect state / history in case events were missed. The tricky thing about this is to beware of side-effects: say you want to get some info about the most recent `sync`. If you activate the daemon for that, it will have the side-effect of also attempting to connect to the network -- which may be pointless, because of the WiFi cable situation. In that sense, journaling event logs to disk as @alexjg  suggested might be worth considering.

Reply:
Yeah this is increasingly looking to me like we need a log and not a pubsub system. With a log we don't need to make any assumptions about what systems are running when producing or consuming events, we don't need to abstract over any platform capabilities, and we can more easily implement transactional semantics where we need to.

The downside is that consumers become more complex, but not enormously so I think, we don't need to build Kafka. Consumers just need to store an  offset and handle log file rotations when reading from the log files.

Reply:
That is phrased _slightly_ on the optimistic side 😅

Reply:
Ha, I suppose it is. I'm not suggesting that it's a trivial endeavour to build a log, just that a) it's not the same challenge as building a distributed log and b) I think that architecturally it concentrates the complexity in one place, rather than having the complexity of handling ephemeral messages and disappaering services spread around the various components of the system.

Reply:
Yeah I’m with you. My concern is just that reliable message buses typically lead to YOLOKAFKA syndrome, meaning that everyone is relying on this component to never lose any data.

I’d favour to first try and _reduce_ the amount of bookkeeping required. For example, once #653 lands, it would not strictly be necessary to remember what went out previously. You can just force a sync with your bootstrap nodes, and let git figure out if any work needs to be performed.

Perhaps we should start with pubsub = `ff01::1`?

Reply:
Thanks for the elaboration and discussion! I think I misinterpreted a previous conversation and was under the assumption the PubSub component was inherently connected to a running daemon, but as you've pointed out that won't be the case.

So in terms of this RFC, it seems like the assumption that will be made is that there will be a PubSub component. If a CLI subcommand deems an event as interesting to track it SHALL publish this event to the PubSub, following the guidelines of the future PubSub RFC. With a caveat being that "interesting" doesn't fall into YOLOKAFKA syndrome -- I'm not entirely sure how we can prevent that but I suppose since it's the consumer's responsibility to sync then they are also responsible for ignoring message. Does that make sense?